### ISSUE TITLE
Ability to access query name string for cache removal

### ISSUE THREAD

**Issue by kvenn** [heftapp/graphql_codegen#173]
If this is possible already, this is mostly a question. If it's not possible, it might be a feature request :) 

I'd like to be able to evict an entire query from cache. To do that, I need to have access to the string version of the query's operation name. 

Something like the following:

```dart
  invalidateCachedQuery(String queryName) {
    final cache = _store.get('Query');
    if (cache == null) return;
    if (cache[queryName] == null) return;
    cache.remove(queryName);
  }
```

I couldn't find anything in the generated code that made this easy, especially when one "query" can actually contain multiple query "operations". The value that's stored in cache is always the query operation name as defined in the schema.

<img width="496" alt="Screen Shot 2022-08-28 at 4 52 10 PM" src="https://user-images.githubusercontent.com/1782467/187094145-e149a80b-897b-4a69-9a39-9bf1abec6199.png">

The only place I've found this from the generated code is deeply nested on the `DocumentNode`

<img width="516" alt="Screen Shot 2022-08-26 at 6 36 22 PM" src="https://user-images.githubusercontent.com/1782467/187094155-8440d1ad-c0bc-491e-930c-783f2b6776cf.png">
 
 In this example, I'd like to extract the string `getState` which requires OperationNode.selectionSet.selections[0].name.value - but this isn't always guaranteed to be in the 0th position.
 
It would be great even if it was a static variable that was exposed that was something like `Operations.<queryName>` that returned the string version. I could hard code this myself, but if the schema changes, I'd have to remember to update this value in multiple locations.

Please let me know if this a bad idea or outside the scope of this project.

Thanks

**Comment by budde377** [comment]
Thanks for reaching out! 

So the operation will always be the first entry on the document node followed by any fragments, but regardless, you can extract the name with something like

```dart
final name = documentNode.definitions.whereType<OperationDefinitionNode>.first.name.value
```

The question is if this is at all useful for cache eviction. Unless you're using some custom cache, most caches will not store queries, but rather objects under some cache key. So you wouldn't be evicting queries as much as you're evicting objects returned by a query. 

In your example, you seem to confuse the operation name with fields on the `Query` object. A query can request multiple fields on the `Query` object per query, but if you really really want to evict these fields, you can do something along the lines of: 

```dart
final fieldForQuery = documentNode.definitions
  .whereType<OperationDefinitionNode>()
  .expand((q) => q.selectionSet.selections.whereType<FieldNode>());
```

but again, evicting these fields will probably not have a huge impact on your cache size, instead, I'd try and evict objects.

**Comment by kvenn** [comment]
This is unbelievably helpful, thank you! I wasn't sure if it was guaranteed the operation was always the first node, so didn't want to assume for a core piece of logic.

In terms of use case (evicting field on Query vs evicting objects), I'm curious if my use case makes sense - but I believe evicting query field would achieve what I want.

* I have a `getFeedItems` field on Query in my schema. I run an operation `query GetFieldItems` when the user goes to their feed page. I run it with a policy of `cacheFirst`.
  ```graphql
  query GetFieldItems {
      getFeedItems {
          ...FeedItem
      }
  }
  ```
* Performing this operation adds a key/value to my HiveStore cache `Query[getFeedItems]`
* The user then does something on another page that would invalidate the object stored in cache (log out, unfriend someone, etc). 
* When the user goes back to view their feed page, with the `cacheFirst` policy, it would return a bunch of invalid FeedItems.

I'd like to tell the GraphQL system that, upon using a `cacheFirst` request to `GetFieldItems`, it should not try to hit cache, and instead go straight to network (whenever a user performs an action that would invalidate the results of that query field).

I believe `refetchQueries` exists for a similar reason, but in my case, I don't think that would be viable (the request is expensive and could introduce a race condition for the user).

**Comment by budde377** [comment]
I think the only current solution is to access to `Store` directly, Hive in your case, and then update the `Query` entry with a version where the fields are removed. 

I think that this issue is a duplicate of #166 from GraphQL Codegen's perspective, and we'll need a change to the GraphQL Flutter API before we can implement this.

**Comment by kvenn** [comment]
Yep that linked enhancement would solve it! 

And using your recommendation of traversing the document nodes should be enough to get the query field in most cases (when it's just a single query field per operation) to then extract from the Hive store. 

Thank you @budde377 🙏
